syntax = "proto3";

package sctl_federated;

// The FederatedLearning service definition.
service FederatedLearning {
  
  // 1. Edge Worker calls this to get the current global model weights (W_t)
  rpc GetGlobalModel (ModelRequest) returns (ModelWeights) {}

  // 2. Edge Worker calls this to send its locally trained weights (W_t^k) back to the server
  rpc SendModelUpdate (ModelUpdate) returns (Acknowledgement) {}

  // 3. Optional: Heartbeat to ensure Zero-Trust connection is alive
  rpc HealthCheck (Ping) returns (Pong) {}
}

// Request for the current global model
message ModelRequest {
  string worker_id = 1;
}

// The core model parameters (weights)
message ModelWeights {
  int32 round_number = 1;
  // We send weights as bytes (serialized dictionary of tensors)
  bytes weights_data = 2; 
}

// The update sent from the worker to the server
message ModelUpdate {
  string worker_id = 1;
  int32 active_worker = 1;
  int32 round_number = 2;
  bytes weights_data = 3;   // The updated weights W_t^k
  int32 num_samples = 4;    // The n_k used for FedAvg weighting
  float anomaly_score = 5;  // Cyber-security metric calculated locally
}

// Simple acknowledgement from the server
message Acknowledgement {
  bool success = 1;
  string message = 2;
}

// Health check messages
message Ping { string client_id = 1; }
message Pong { string status = 1; }